extract.go

```diff
@@ -207,6 +208,113 @@ func (e *Extractor) UseEmbeddedStructs(use bool) *Extractor {
 	return e
 }
 
+// GetChangesetForTag takes a dataset and tries to change the keys from one
+// tag to another. This is helpful when trying to create partial updates to a
+// row in a database.
+func (e *Extractor) GetChangesetForTag(data map[string]interface{}, inputTag, outputTag string) (out map[string]interface{}, err error) {
+	if err = e.isValidStruct(); err != nil {
+		return
+	}
+
+	out = make(map[string]interface{})
+	s := reflect.ValueOf(e.StructAddr).Elem()
+	fields := e.fields(s)
+	for _, field := range fields {
+
+		inputFieldTag, ok := field.tags.Lookup(inputTag)
+		if !ok {
+			continue
+		}
+		outputFieldTag, ok := field.tags.Lookup(outputTag)
+		if !ok {
+			continue
+		}
+		value, ok := data[inputFieldTag]
+		if !ok {
+			continue
+		}
+		out[outputFieldTag] = value
+	}
+
+	return
+}
+
+// ApplyMap takes a map of arbitrary values and tries to apply them to a
+// struct, matching keys with the given input tag. If the input tag is empty,
+// the key will be matched against the field name
+func (e *Extractor) ApplyMap(data map[string]interface{}, inputTag string) (interface{}, error) {
+	err := e.isValidStruct()
+	if err != nil {
+		return nil, err
+	}
+
+	validFields := map[string]bool{}
+	s := reflect.ValueOf(e.StructAddr).Elem()
+	fields := e.fields(s)
+	for _, field := range fields {
+
+		var inputFieldTag string
+		var ok bool
+		if inputTag == "" {
+			inputFieldTag = field.name
+		} else {
+			inputFieldTag, ok = field.tags.Lookup(inputTag)
+			if !ok {
+				continue
+			}
+		}
+
+		_, ok = data[inputFieldTag]
+		if !ok {
+			continue
+		}
+
+		fieldName := field.name
+		validFields[fieldName] = true
+
+	}
+
+	// Create a new zero value of the type stored in e.StructAddr
+	out := reflect.New(s.Type()).Interface()
+	bytes, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	// dirty hax, as some fields may be deserialized differently depending
+	// on whether they're being deserialized into a map[string]interface{}
+	// or a user specified type.
+	err = json.Unmarshal(bytes, out)
```

Comment:
💣

Reply:
I'm ok with the PR being declined based on the dirty hack. I'm also open to suggestions on how to do it in a less hacky way.